File: answers/25/main.py
import numpy as np
import cv2
import matplotlib.pyplot as plt
import sys
from answers.hist import *


def nn_inter(img, a, b):
    out_h = int(img.shape[0] * a)
    out_w = int(img.shape[1] * b)
    out = np.zeros([out_h, out_w, img.shape[2]], dtype=np.uint8)

    xs, ys = np.meshgrid(range(out_h), range(out_w))
    out[ys, xs] = img[np.round(ys / a).astype(int), np.round(xs / b).astype(int)]
    return out

if __name__ == '__main__':
    img_orig = cv2.imread('questions/dataset/images/imori_256x256.png')
    img_result = nn_inter(img_orig, a=1.5, b=1.5)

    cv2.imshow('python_origin', img_orig)
    cv2.imshow('python_result', img_result)
    cv2.moveWindow('python_result', img_orig.shape[1], 0)
    cv2.waitKey(0)

File: answers/25/main.cpp
#include <vector>
#include <chrono>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <unistd.h>
#include <cmath>
#include "../hist.cpp"

void nn_inter(const cv::Mat &src, cv::Mat &dst, float a, float b) {
    // 出力画像サイズ
    int dst_h = src.size().height * a;
    int dst_w = src.size().width * b;
    dst = cv::Mat::zeros(cv::Size(dst_h, dst_w), CV_8UC3);

    dst.forEach<cv::Vec3b>([&](cv::Vec3b &pixel, const int pos[]) -> void {
        int x = pos[1];
        int y = pos[0];
        //pixel = src.at<cv::Vec3b>(int(y / a), int(x / b));
        pixel = src.at<cv::Vec3b>(round(y / a), round(x / b));
    });
}

int main() {
    cv::Mat img_orig = cv::imread("questions/dataset/images/imori_256x256.png");
    cv::Mat img_result;
    nn_inter(img_orig, img_result, 1.5, 1.5);

    cv::imshow("cpp_origin", img_orig);
    cv::imshow("cpp_result", img_result);
    cv::moveWindow("cpp_origin", 0, -img_orig.size().height);
    cv::moveWindow("cpp_result", img_orig.size().width, -img_orig.size().height);
    cv::waitKey(0);
    return 0;
}
